### ISSUE TITLE
Quantized CLIPModel inference not noticeably faster (or even slower) than non quantized

### ISSUE THREAD

**Issue by kechan** [huggingface/quanto#202]
I don't have exhaustive timing. Here's my code snippet:

```
model = CLIPModel.from_pretrained("laion/CLIP-ViT-L-14-laion2B-s32B-b82K")
processor = CLIPProcessor.from_pretrained("laion/CLIP-ViT-L-14-laion2B-s32B-b82K")

inputs = processor(text=texts, images=[image], return_tensors="pt", padding=True)

quantize(model, weights=qint8, activations=None)
freeze(model)
model.eval();

start_time = time.time()
with torch.no_grad():  
    outputs = model(**inputs)
    text_embeddings = outputs.text_embeds  
    image_embeddings = outputs.image_embeds  

end_time = time.time()  # End timing
print(f"Execution time: {end_time - start_time} seconds")  
```

On colab without GPU, this runs to be about ~4s, while without quantized (the original) runs at ~3s. Wonder if I did something wrong. I shared the notebook here with the reproducing code:

https://colab.research.google.com/drive/1qOG6WtCyaO1vzSK4ZWieKPJGZAItJ570?usp=sharing

**Comment by dacorvo** [comment]
@kechan this is perfectly normal. Without dedicated hardware support for matrix multiplications involving lower bitwidth tensors, the inference is slower because of the extra operations to "dequantize" and obtain float tensors before the operation.

**Comment by kechan** [comment]
This is rather disappointing. The reason I want to quantize is for both lower memory footprint and faster inference.  if this issue is "normal" across the board, I wonder why people don't stress this. This has become completely impractical for me for deployment, where my original model is slow enough.

**Comment by dacorvo** [comment]
You can try `qint4` on GPU, whose performance is closer to `float16`. 
For a very short and simplified explanation of why it is slower:
- for efficiency, the hardware (even cpu) only process chunks of the whole tensors at a time, but in parallel in different threads,
- when you use a lower bitwidth, you can process more items per thread because you can fit more of them in each chunk  (typically x2 for 8-bit, and x4 for 4-bit),
- you are however constrained by the arithmetics supported by the hardware, and for instance when multiplying a float input by an integer weight, at some point you need to upcast, which is costly,
- after processing each chunk you need also to upscale it to the scale before quantization, which is also costly.
So one one hand there is a gain, but on the other hand there is a loss.
In `quanto`:
- there is not (yet) any optimization for `qint8`, so there is nothing to counterbalance the cost of cast + upscale,
- there is an optimization for `qint4` on CUDA, which allows to almost counterbalance the dequantization cost.

**Comment by kechan** [comment]
Thanks a lot for the explanation, guess have to read and learn more about this. merely matching unquantized inference speed may not be enough for my case, or I am overlooking something. I have some very strict latency requirement, more so than memory constraint, and I am looking for way to use the same model without going for a smaller version in the family, or started going to distillation and or that. 

it probably looks like quantization is not an "automatic" solution to speed things up, too many things to consider, beside evaluating for the sacrifice in accuracy performance.

**Comment by kechan** [comment]
anyway, I actually tried qint4 on google colab with a T4 GPU, it failed to generate inference completely (ending up with 0s in the embedding). here's the colab:

https://colab.research.google.com/drive/1qOG6WtCyaO1vzSK4ZWieKPJGZAItJ570?usp=sharing

/usr/local/lib/python3.10/dist-packages/quanto/library/ext/cuda/awq/v2/gemm_cuda.cu(300): warning #177-D: variable "blockIdx_x" was declared but never referenced
    int blockIdx_x = 0;
        ^

Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"

/usr/local/lib/python3.10/dist-packages/quanto/library/ext/cuda/awq/v2/gemm_cuda.cu(320): warning #177-D: variable "kSmemSizeZeros" was declared but never referenced
    constexpr int kSmemSizeZeros = CTA_N * STAGES / scales_load_interval * scales_per_load;

and tons and tons of error messages.

**Comment by kechan** [comment]
I mean those are warnings, these are the real errors:

ptxas /tmp/tmpxft_00001b14_00000000-6_gemm_cuda.ptx, line 146; error   : Feature 'cp.async' requires .target sm_80 or higher
ptxas /tmp/tmpxft_00001b14_00000000-6_gemm_cuda.ptx, line 161; error   : Feature 'cp.async' requires .target sm_80 or higher
ptxas /tmp/tmpxft_00001b14_00000000-6_gemm_cuda.ptx, line 188; error   : Feature 'cp.async' requires .target sm_80 or higher
ptxas /tmp/tmpxft_00001b14_00000000-6_gemm_cuda.ptx, line 213; error   : Feature 'cp.async' requires .target sm_80 or higher
ptxas /tmp/tmpxft_00001b14_00000000-6_gemm_cuda.ptx, line 220; error   : Feature 'cp.async' requires .target sm_80 or higher
ptxas /

**Comment by dacorvo** [comment]
Oh right: this does not work on earlier generations of GPUs. I have created an issue to at least detect that early on: #206

**Comment by kechan** [comment]
Please see my update. it may have worked, but haven't fully evaluate correctness/accuracy. but inference perf is still bad.

**Comment by dacorvo** [comment]
The compilation of the CUDA kernels worked, but their execution fails, which means that your linear layers will launch the kernels, get an exception then fallback to pytorch _every_single_time__. This will be really slow. I will improve that by detecting early on that the GPU is too old.
